tck/src/main/java/org/reactivestreams/tck/TestEnvironment.java

```diff
@@ -50,15 +50,47 @@
    * the implementation, but can in some cases result in longer time to
    * run the tests.
    * @param defaultTimeoutMillis default timeout to be used in all expect* methods
+   * @param defaultPollTimeoutMillis default amount of time to poll for events if {@code defaultTimeoutMillis} isn't
+   *                                 preempted by an asynchronous event.
    * @param defaultNoSignalsTimeoutMillis default timeout to be used when no further signals are expected anymore
    * @param printlnDebug         if true, signals such as OnNext / Request / OnComplete etc will be printed to standard output,
    */
-  public TestEnvironment(long defaultTimeoutMillis, long defaultNoSignalsTimeoutMillis, boolean printlnDebug) {
+  public TestEnvironment(long defaultTimeoutMillis, long defaultNoSignalsTimeoutMillis, long defaultPollTimeoutMillis,
+                         boolean printlnDebug) {
     this.defaultTimeoutMillis = defaultTimeoutMillis;
+    this.defaultPollTimeoutMillis = defaultPollTimeoutMillis;
     this.defaultNoSignalsTimeoutMillis = defaultNoSignalsTimeoutMillis;
     this.printlnDebug = printlnDebug;
   }
 
+  /**
+   * Tests must specify the timeout for expected outcome of asynchronous
+   * interactions. Longer timeout does not invalidate the correctness of
+   * the implementation, but can in some cases result in longer time to
+   * run the tests.
+   * @param defaultTimeoutMillis default timeout to be used in all expect* methods
+   * @param defaultNoSignalsTimeoutMillis default timeout to be used when no further signals are expected anymore
+   * @param printlnDebug         if true, signals such as OnNext / Request / OnComplete etc will be printed to standard output,
+   */
+  public TestEnvironment(long defaultTimeoutMillis, long defaultNoSignalsTimeoutMillis, boolean printlnDebug) {
+    this(defaultTimeoutMillis, defaultNoSignalsTimeoutMillis, defaultTimeoutMillis, printlnDebug);
+  }
+
+  /**
+   * Tests must specify the timeout for expected outcome of asynchronous
+   * interactions. Longer timeout does not invalidate the correctness of
+   * the implementation, but can in some cases result in longer time to
+   * run the tests.
+   *
+   * @param defaultTimeoutMillis default timeout to be used in all expect* methods
+   * @param defaultNoSignalsTimeoutMillis default timeout to be used when no further signals are expected anymore
+   * @param defaultPollTimeoutMillis default amount of time to poll for events if {@code defaultTimeoutMillis} isn't
+   *                                 preempted by an asynchronous event.
+   */
+  public TestEnvironment(long defaultTimeoutMillis, long defaultNoSignalsTimeoutMillis, long defaultPollTimeoutMillis) {
+      this(defaultTimeoutMillis, defaultNoSignalsTimeoutMillis, defaultPollTimeoutMillis, false);
+  }
```

Comment:
the constructor dance is not yet pushed all the way; we have to implement being able to pick up the value from ENV, as people sometimes rely on doing this on CI etc.

So you'll need 

```java
  /**
   * Tries to parse the env variable {@code DEFAULT_POLL_TIMEOUT_MILLIS_ENV} as long and returns the value if present OR its default value.
   *
   * @throws java.lang.IllegalArgumentException when unable to parse the env variable
   */
  public static long envDefaultPollTimeoutMillis() {
    final String envMillis = System.getenv(DEFAULT_POLL_TIMEOUT_MILLIS_ENV);
    if (envMillis == null) return envDefaultTimeoutMillis();
    else try {
      return Long.parseLong(envMillis);
    } catch (NumberFormatException ex) {
      throw new IllegalArgumentException(String.format("Unable to parse %s env value [%s] as long!", DEFAULT_POLL_TIMEOUT_MILLIS_ENV, envMillis), ex);
    }
  }

```

and use it in:

```java
  public TestEnvironment(boolean printlnDebug) {
    this(envDefaultTimeoutMillis(), envDefaultPollTimeoutMillis(), envDefaultNoSignalsTimeoutMillis(), printlnDebug);
  }
```

And also

```java
  public TestEnvironment(long defaultTimeoutMillis, long defaultNoSignalsTimeoutMillis, boolean printlnDebug) {
    this(defaultTimeoutMillis, envDefaultPollTimeoutMillis(), defaultNoSignalsTimeoutMillis, printlnDebug);
  }
```

(yeah I know it's such a pain... wish Java would gain default params already :P)

Let's also reorder the constructor to be:

```java
  public TestEnvironment(long defaultTimeoutMillis, 
                         long defaultPollTimeoutMillis,
                         long defaultNoSignalsTimeoutMillis, 
                         boolean printlnDebug) {
```

sine the poll is more related to the timeout than the no signals I think (order of params must be the same _everywhere_, declarations, docs, fields).

Reply:
> And also...

The intention was to preserve backwards compatibility/behavior in this PR. If we grab the value from the environment here we will no longer be preserving existing behavior. Are you sure you want

> Let's also reorder the constructor to be:

The relative ordering of arguments would be different than other constructors.

This already exists...
```java
public TestEnvironment(long defaultTimeoutMillis, long defaultNoSignalsTimeoutMillis)
```

so we cannot have:
```java
public TestEnvironment(long defaultTimeoutMillis, long defaultPollTimeoutMillis)
```

This would make for a subtle issue in transitioning to the following constructor (the types are compatible but the argument order/semantics swap):
```java
public TestEnvironment(long defaultTimeoutMillis, long defaultNoSignalsTimeoutMillis, long defaultPollTimeoutMillis)
```

I can re-order if you prefer but not sure if this is a clear win (considering the existing api/behavior).

Reply:
Hah, another good point about the compat here... Ignore me here, was looking at it from a too "api puristic" point of view there.

Reply:
ok I will keep as is. fwiw I agree your suggestion is an improvement from an API perspective 👍